src/util/timelock.rs

```diff
@@ -0,0 +1,236 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Bitcoin timelocks.
+//!
+//! Relative and absolute timelocks based on `nSequence` and `nLockTime`.
+//!
+
+#[cfg(feature = "std")]
+use std::error;
+
+use core::fmt;
+
+use blockdata::constants::{
+    LOCKTIME_THRESHOLD, SEQUENCE_LOCKTIME_DISABLE_FLAG, SEQUENCE_LOCKTIME_GRANULARITY,
+    SEQUENCE_LOCKTIME_TYPE_FLAG,
+};
+
+/// An absolute time lock, expires after either height or time.
+///
+/// This is nLockTime as defined in BIP-65.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Abs(u32);
+
+impl Abs {
+    /// Returns true if this absolute lock is lock-by-blockheight.
+    pub fn is_block_time(self) -> bool {
+        self.0 < LOCKTIME_THRESHOLD
+    }
+
+    /// Returns true if this absolute lock is lock-by-blocktime.
+    pub fn is_block_height(self) -> bool {
+        self.0 >= LOCKTIME_THRESHOLD
+    }
+
+    /// Returns true if this timelock is the same type as `other` timelock.
+    pub fn is_same_type(&self, other: Self) -> bool {
+        self.is_block_time() && other.is_block_time()
+            || self.is_block_height() && other.is_block_height()
+    }
+
+    /// Returns true if the timelock is locked at `age`.
+    pub fn is_locked(&self, age: u32) -> Result<bool, Error> {
+        self.is_expired(age).map(|b| !b)
+    }
+
+    /// Returns true if the timelock has expired at `age`.
+    pub fn is_expired(&self, age: u32) -> Result<bool, Error> {
+        let age = Abs::from(age);
+        if !self.is_same_type(age) {
+            return Err(Error::InvalidComparison);
+        }
+        Ok(age.0 >= self.0)
+    }
+
+    /// Returns true if this timelock value is zero.
```

Comment:
Does it have any specific meaning?

Reply:
In theory 0 is no different from any other value. In practice a relative timelock of zero means that you can chain unconfirmed transactions off of each other, while an absolute timelock of zero means "no timelock". So there's a qualitative difference which I can imagine would be useful to check for.

Reply:
My point was mostly that it would be good to describe what "equal to 0" means and why would it be useful to call, for educational purposes. :)